File: src/main/java/com/linebot/action/ActionHandler.java
package com.linebot.action;

import com.linebot.message.FlexMessageBuilder;
import com.linebot.model.UserStatus;
import com.linebot.service.UserStatusCacheService;
import com.linebot.service.log.LogService;
import com.linebot.service.notice.NoticeService;
import com.linebot.service.user.BotUserQiitaService;
import com.linebot.service.user.BotUserService;
import com.linecorp.bot.model.message.Message;
import com.linecorp.bot.model.message.TextMessage;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.ApplicationContext;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import javax.annotation.Nullable;
import javax.validation.constraints.NotNull;
import java.util.Arrays;
import java.util.List;

@Slf4j
@AllArgsConstructor
@Component
public class ActionHandler {
    private ApplicationContext applicationContext;
    private UserStatusCacheService userStatusCacheService;
    private BotUserService botUserService;
    private BotUserQiitaService botUserQiitaService;
    private NoticeService noticeService;
    private FlexMessageBuilder flexMessageBuilder;
    private LogService logService;

    public List<Message> follow(@NotNull String userId) {
        botUserService.insert(userId);
        return Arrays.asList(
                new TextMessage("ユーザ登録を行いました。以下操作が実行可能です。"),
                flexMessageBuilder.buildStartWordMessage()
        );
    }

    @Transactional
    public void unfollow(@NotNull String userId) {
        botUserService.delete(userId);
        botUserQiitaService.delete(userId);
        noticeService.deleteNotice(userId);
    }

    @NotNull
    public List<Message> handle(@NotNull String userId, @NotNull String message) {
        ActionSelector actionSelector = ActionSelector.getByStartWord(message);
        if (actionSelector != null) {
            return executeStartAction(actionSelector, userId, message);
        }

        UserStatus status = this.getStatus(userId);
        if (status.getNextAction() == null) {
            return Arrays.asList(
                    new TextMessage("このメッセージは受け付けられません。どの操作を実行するか選択してください"),
                    flexMessageBuilder.buildStartWordMessage()
            );
        }

        Action action = getAction(status.getNextAction());
        if (action.check(message)) {
            List<Message> result = execute(action, userId, message);
            userStatusCacheService.set(userId, createUserStatus(userId, action.getNextAction()));
            return result;
        } else {
            userStatusCacheService.delete(userId);
            return Arrays.asList(
                    new TextMessage("このメッセージは受け付けられません。どの操作を実行するか選択してください"),
                    flexMessageBuilder.buildStartWordMessage()
            );
        }
    }

    private List<Message> execute(@NotNull Action action, @NotNull String userId, @NotNull String message) {
        logService.insertBotLog(userId, action.getClass().getSimpleName(), message);
        return action.execute(userId, message);
    }

    private List<Message> executeStartAction(ActionSelector actionSelector, String userId, String message) {
        Action action = getStartAction(actionSelector);
        // スタートワードなのでチェック不要でexecute実行する
        List<Message> result = execute(action, userId, message);
        userStatusCacheService.set(userId, createUserStatus(userId, action.getNextAction()));
        return result;
    }

    private Action getStartAction(ActionSelector actionSelector) {
        String actionName = actionSelector.getActionList().stream()
                .findFirst()
                .get();
        return getAction(actionName);
    }

    private Action getAction(String actionName) {
        return (Action) applicationContext.getBean(actionName);
    }

    private UserStatus getStatus(String userId) {
        UserStatus u = userStatusCacheService.get(userId);
        return u != null ? u : new UserStatus();
    }

    private UserStatus createUserStatus(@NotNull String userId, @Nullable String nextAction) {
        UserStatus u = new UserStatus();
        u.setUserId(userId);
        u.setNextAction(nextAction);
        return u;
    }
}

File: src/main/kotlin/com/linebot/action/ActionHandler.kt
package com.linebot.action

import com.linebot.message.FlexMessageBuilder
import com.linebot.model.UserStatus
import com.linebot.service.cache.UserStatusCacheService
import com.linebot.service.notice.NoticeService
import com.linebot.service.user.BotUserService
import com.linecorp.bot.model.message.Message
import com.linecorp.bot.model.message.TextMessage
import org.springframework.context.ApplicationContext
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class ActionHandler(
    private val applicationContext: ApplicationContext,
    private val userStatusCacheService: UserStatusCacheService,
    private val botUserService: BotUserService,
    private val noticeService: NoticeService,
    private val flexMessageBuilder: FlexMessageBuilder
) {

    fun follow(userId: String): List<Message> {
        botUserService.insert(userId)
        return listOf(
                TextMessage("ユーザ登録を行いました。以下操作が実行可能です。"),
                flexMessageBuilder.buildStartWordMessage()
        )
    }

    @Transactional
    fun unfollow(userId: String) {
        botUserService.delete(userId)
        noticeService.deleteNotice(userId)
    }

    fun handle(userId: String, message: String): List<Message> {
        val actionSelector = ActionSelector.getByStartWord(message)
        if (actionSelector != null) {
            return executeStartAction(actionSelector, userId, message)
        }

        return getStatus(userId).nextAction
                ?.let { getAction(it) }
                ?.takeIf { it.check(message) }
                ?.let {
                    it.execute(userId, message).also { _ ->
                        userStatusCacheService.set(userId, UserStatus(userId, it.nextAction))
                    }
                } ?: return handleError(userId)
    }

    private fun executeStartAction(actionSelector: ActionSelector, userId: String, message: String): List<Message> {
        val action = getStartAction(actionSelector)
        // スタートワードなのでチェック不要でexecute実行する
        val result = action.execute(userId, message)
        userStatusCacheService.set(userId, UserStatus(userId, action.nextAction))
        return result
    }

    private fun handleError(userId: String): List<Message> {
        userStatusCacheService.delete(userId)
        return listOf(
                TextMessage("このメッセージは受け付けられません。どの操作を実行するか選択してください"),
                flexMessageBuilder.buildStartWordMessage()
        )
    }

    private fun getStartAction(actionSelector: ActionSelector) = getAction(actionSelector.actionList.first())
    private fun getAction(actionName: String) = applicationContext.getBean(actionName) as Action
    private fun getStatus(userId: String) = userStatusCacheService.get(userId) ?: UserStatus()
}
